extensions/wikia/Chat2/js/emoticons.js

```diff
@@ -11,99 +11,138 @@
-	for(var regexString in imgUrlsByRegexString){
-		/*
-		 * empty string for emote icons crash Chat
-		 * so ignore them
-		 */
-		if(regexString == '') continue;
+WikiaEmoticons.doReplacements = function (text, emoticonMapping) {
+	// This debug is probably noisy, but I added it here just in case all of these regexes turn out
+	// to be slow (so we could see that in the log & know that we need to make this function more efficient).
+	$().log('Processing any emoticons... ');
+
+	var imgUrlsByRegexString = emoticonMapping.getImgUrlsByRegexString(),
+		regexString,
+		imgSrc,
+		numIters,
+		origText,
+		regex,
+		glyphUsed,
+		buildTagFunc;
+
+	for (regexString in imgUrlsByRegexString){
+		// Empty string for emote icons crash Chat so ignore them
+		if (regexString === '') {
+			continue;
+		}
 
 		imgSrc = imgUrlsByRegexString[regexString];
-		imgSrc = imgSrc.replace(/"/g, "%22"); // prevent any HTML-injection
+		imgSrc = imgSrc.replace(/"/g, '%22'); // prevent any HTML-injection
+
+		buildTagFunc = WikiaEmoticons.buildTagGenerator(imgSrc);
 
 		// Fix > and <
-		regexString = regexString.replace(/>/g, "&gt;");
-		regexString = regexString.replace(/</g, "&lt;");
-
-		// Build the regex for the character (make it ignore the match if there is a "/" immediately after the emoticon. That creates all kinds of problems with URLs).
-		var numIters = 0;
-		var origText = text;
-		do{
-			var regex = new RegExp("(^| )(" + regexString + ")([^/]|$)", "gi"); // NOTE: \s does not work for whitespace here for some reason.
-			var emoticon = " <img src=\""+imgSrc+"\" width='"+WikiaEmoticons.EMOTICON_WIDTH+"' height='"+WikiaEmoticons.EMOTICON_HEIGHT+"' alt=\"$2\" title=\"$2\"/> ";
-			var glyphUsed = text.replace(regex, '$2');
-			glyphUsed = glyphUsed.replace(/"/g, "&quot;"); // prevent any HTML-injection
-			text = text.replace(regex, '$1' + emoticon + '$3');
-		} while ((origText != text) && (numIters++ < 5));
+		regexString = regexString.replace(/>/g, '&gt;');
+		regexString = regexString.replace(/</g, '&lt;');
+
+		// Build the regex for the character (make it ignore the match if there is a "/" immediately
+		// after the emoticon. That creates all kinds of problems with URLs).
+		numIters = 0;
+		origText = text;
+
+		do {
+			// NOTE: \s does not work for whitespace here for some reason.
+			regex = new RegExp('(^| )(' + regexString + ')([^/]|$)', 'gi');
+			glyphUsed = text.replace(regex, '$2');
+			glyphUsed = glyphUsed.replace(/"/g, '&quot;'); // prevent any HTML-injection
+			text = text.replace(regex, buildTagFunc);
+		} while ((origText !== text) && (numIters++ < 5));
 	}
 
-	$().log("Done processing emoticons.");
+	$().log('Done processing emoticons.');
 	return text;
-} // end doReplacements()
+}; // end doReplacements()
+
+WikiaEmoticons.buildTagGenerator = function (imgSrc) {
+	return function (match, leading, tag, trailing) {
+		tag = mw.html.escape(tag);
+		return (
+			leading +
+			' <img ' +
+				'src="' + imgSrc + '" ' +
```

Comment:
Actually, I forgot that part of the acceptance criteria was making sure the image source was from Wikia.  I'll do both of those ...